index.bs

```diff
@@ -2039,24 +2139,53 @@ input parameters:
     by the client), if any.
+: |requireUserVerification|
+:: The [=effective user verification requirement for assertion=], a Boolean value provided by the client.
 : |extensions|
 :: A [=map=] from [=extension identifiers=] to their [=authenticator extension inputs=], created by the client based on the
     extensions requested by the [=[RP]=], if any.
 
 When this method is invoked, the [=authenticator=] must perform the following procedure:
 1. Check if all the supplied parameters are syntactically well-formed and of the correct length. If not, return an error code
     equivalent to "{{UnknownError}}" and terminate the operation.
+1. Let |isFirstFactorOperation| be a Boolean value, set to `true` if and only if |allowCredentialDescriptorList| was not supplied.
+1. If at least one of |requireUserVerification| and |isFirstFactorOperation| is `true` and the authenticator cannot perform [=user
+    verification=], return an error code equivalent to "{{ConstraintError}}" and terminate the operation.
 1. If |allowCredentialDescriptorList| was not supplied, set it to a list of all credentials stored for |rpId| (as determined by
     an exact match of |rpId|).
-1. Remove any items from |allowCredentialDescriptorList| that do not match a credential bound to this authenticator. A match 
-    occurs if a credential matches <code>|rpId|</code> and an |allowCredentialDescriptorList| item's <code>|allowCredentialDescriptorList| 
-    {{PublicKeyCredentialDescriptor/id}}</code> and <code>|allowCredentialDescriptorList|{{PublicKeyCredentialDescriptor/type}}</code>. 
+1. Remove any items from |allowCredentialDescriptorList| that do not match a credential bound to this authenticator. A match
+    occurs if a credential matches <code>|rpId|</code> and an |allowCredentialDescriptorList| item's
+    {{PublicKeyCredentialDescriptor/id}} and {{PublicKeyCredentialDescriptor/type}} members.
 1. If |allowCredentialDescriptorList| is now empty, return an error code equivalent to "{{NotAllowedError}}" and terminate the
     operation.
+
+1. If |isFirstFactorOperation| is `true`, perform [=user verification=]. If [=user verification=] fails, return an error code
+    equivalent to "{{NotAllowedError}}" and terminate the operation.
+
+    Remove any items from |allowCredentialDescriptorList| that match a credential that was created by a [=user verified|verified
+    user=] other than the current [=user verified|verified user=].
+
 1. Prompt the user to select a [=public key credential|credential=] |selectedCredential| from |allowCredentialDescriptorList|.
-    Obtain [=user consent=] for using |selectedCredential|. The prompt for obtaining this [=user consent|consent=] may be shown
-    by the
-    [=authenticator=] if it has its own output capability, or by the user agent otherwise.
+
+1. Obtain [=user consent=] for using |selectedCredential|. The prompt for obtaining this [=user consent|consent=] may be shown by
```

Comment:
The spec's original intent is that the act of user verification is the act of obtaining user consent, so splitting them apart here seems incorrect to me.

Reply:
In Client PIN case, user verification is not same as user presence check. User still has to do perform touch or equivalent. This looks correct to me.

Reply:
That was my intent as well, I've tried to formulate that UV and UP are interchangeable as user consent unless UV is specifically required.

My thinking in the case of `authenticatorGetAssertion` was also that there are two scenarios:

- This is a second factor operation (`allowCredentials` given); the user has not been verified before selecting a credential. In this case UV or UP can be done interchangeably in one step to confirm the selection.
- This is a first factor operation (`allowCredentials` not given); the user has already been verified before selecting a credential. However it's probably wise to collect user consent to confirm the selection anyway - for example, the authenticator may have skipped the selection prompt if there was only one eligible credential.

Reply:
Hm, I'm not sure I agree with the above comments -- for several reasons...

1. @akshayku wrote:
    > In Client PIN case, user verification is not same as user presence check. 

    I do not see that stated in the [CTAP spec wrt Client PIN](https://fidoalliance.org/specs/fido-v2.0-rd-20170927/fido-client-to-authenticator-protocol-v2.0-rd-20170927.html#authenticatorClientPIN). 

    If I recall correctly, you argued that a user's entering a Client PIN on an authnr supporting Client PIN ought to interpreted as an act of [User Verification](https://w3c.github.io/webauthn/#user-verification).  I do not understand how that can not also be interpreted as a [user presence test](https://w3c.github.io/webauthn/#test-of-user-presence). 

2. @emlun wrote: 
    > This is a first factor operation (allowCredentials not given); the user has already been verified before selecting a credential.

    I am not sure that the absence of `allowCredentials` necessarily denotes a "first factor" operation (aka "passwordless").  What do you mean by "the user has already been verified before selecting a credential" ?  Do you mean that the user was "verified" by presence of a cookie, say?

    If I understand correctly (IIUC), in the case where the browser has valid ambient authentication credentials, e.g. cookies, for the RP, then the RP will be able (and ought) to supply `allowCredentials` because the RP will know who the user is.  I.e., this authn aka getAssertion operation is being a form a "step-up" in that the RP knows who the nominally-authn'd user is, and is employing webauthn to ensure that.  

    And IIUC, this is regardless of whether the authnr is performing as 1st multi-factor or 2nd factor.  @rlin1, what do you think ?

3. Does an authnr operating over CTAP perform the same detailed steps as are now specified in [WebAuthn's Authenticator model's operations](https://w3c.github.io/webauthn/#authenticator-model) ?

4. Overall, these autenticatorMakeCredential  and  authenticatorGetAssertion operations are _supposed_ to be "fuzzy" and leave room for platform implementors to figure out how implement this layer. CTAP is only one concrete instantiation of these operations.  Also, one can argue that there is an unspecified-in-webauthn layer between the [WebAuthn Authenticator model](https://w3c.github.io/webauthn/#authenticator-model) (of which the above operations are only a portion). For example, CTAP's client pin seems to occupy this unspecified-in-webauthn layer.  Are the algorithm descriptions becoming too detailed?  Do they match up with the [CTAP spec](https://fidoalliance.org/specs/fido-v2.0-rd-20170927/fido-client-to-authenticator-protocol-v2.0-rd-20170927.html)?

Reply:
(2): What I mean by "first factor operation" is exactly that `allowCredentials` has not been supplied to `credentials.get()`. Unfortunately this currently doesn't exactly correspond to when no `allowCredentialDescriptorList` is supplied to [authenticatorGetAssertion][authnr-get] (see #685), but I'm hoping that will get addressed.

What I mean by "the user has already been verified" is that with these proposed changes, the user has already been verified before step 9 in [authenticatorGetAssertion][authnr-get]. I meant my previous comment to be in the context of step 9 only.

(3-4): Hm, you're right that CTAP doesn't explicitly specify that the user being verified for the assertion must be the same user that created the credential. It looks like CTAP's PIN mechanism is implicitly single-user only, but the same cannot be said for the unspecified other kinds of UV gestures. CTAP also doesn't automatically require UV if `allowList` is not specified.

So I guess I'll have to revert those two things, then.

I agree that the authenticator operations in WebAuthn should be kept as loosely specified as possible, but I thought it seemed like the above were agreed-upon limitations necessary to ensure the user's privacy.

[authnr-get]: https://s3.amazonaws.com/pr-preview/w3c/webauthn/issue-644.html#op-get-assertion

Reply:
We SHOULD be clear that user verification, e.g. a fingerprint swipe IS a method for obtaining user consent - we don't want additional dialog boxes!

Reply:
>We SHOULD NOT require additional user consent gathering by the platform for using an authenticator that is already registered!

It seems like you're saying that `getAssertion` operations should never ask for user consent, and should always be silent. Is that what you mean?

>We SHOULD be clear that user verification, e.g. a fingerprint swipe IS a method for obtaining user consent - we don't want additional dialog boxes!

Does the proposed language below need to be clearer on this point? It mirrors the corresponding part of the `getAssertion` operation in the CTAP2 spec:

>Collect user consent if required. [...]
>- If the "uv" option was specified and set to true:
>   - If device doesn't support user-identifiable gestures, return the CTAP2_ERR_OPTION_NOT_SUPPORTED error.
>   - Collect a user-identifiable gesture. If gesture validation fails, return the CTAP2_ERR_OPERATION_DENIED error.
>- If the "up" option was specified and set to true, collect the user's consent.
>   - If no consent is obtained and a timeout occurs, return the CTAP2_ERR_OPERATION_DENIED error.